feat: Add PURL (Package URL) builtin functions

anivar opened the pull request:
## Summary

This PR implements issue #6504 by adding two new builtin functions for working with Package URLs (PURLs):
- `purl.is_valid(purl)` - Validates that a string is a valid PURL
- `purl.parse(purl)` - Parses a PURL into its component parts

## Motivation

As mentioned in #6504 by @lcarva from the Enterprise Contract project, PURL support is essential for validating Software Bill of Materials (SBOM) entries and implementing supply chain security policies. PURLs are used extensively in CycloneDX and other SBOM formats.

## Implementation

The implementation uses the standard `packageurl-go` library (v0.1.3) to ensure full compatibility with the PURL specification.

### New Functions

**`purl.is_valid(purl)`**
- Returns `true` if the input is a valid PURL, `false` otherwise
- Never throws errors (follows the pattern of `json.is_valid`)

**`purl.parse(purl)`**
- Parses a PURL and returns an object with components
- Throws an error for invalid PURLs

### Example Usage

```rego
# Validate a PURL
valid := purl.is_valid("pkg:npm/express@4.18.2")

# Parse and check components
package_info := purl.parse("pkg:maven/org.apache.xmlgraphics/batik-anim@1.9.1")
allowed := package_info.type == "maven" and package_info.namespace == "org.apache.xmlgraphics"

# Check for vulnerabilities by version
vulnerable := package_info.name == "log4j" and package_info.version < "2.17.0"

# Extract qualifiers
parsed := purl.parse("pkg:rpm/fedora/curl@7.50.3?arch=x86_64")
arch := parsed.qualifiers.arch
```

### Parsed Object Structure

```json
{
  "type": "npm",              // Package type (required)
  "namespace": "org.apache",  // Namespace/group (optional, empty string if absent)
  "name": "commons-lang",     // Package name (required)
  "version": "1.2.3",        // Version (optional, empty string if absent)
  "qualifiers": {            // Qualifiers (always an object, empty {} if none)
    "arch": "x86_64"
  },
  "subpath": "lib/main.js"   // Subpath (optional, empty string if absent)
}
```

## Testing

Added comprehensive unit tests covering:
- Valid PURLs for all common package types (npm, maven, pypi, docker, etc.)
- Invalid PURL detection
- Qualifier and subpath parsing
- Edge cases and error handling

## Impact

This enables OPA users to:
- Validate PURL entries in SBOMs
- Extract package metadata for policy decisions
- Implement supply chain security policies
- Detect disallowed software components

Fixes #6504

cc @lcarva @ashutosh-narkar

srenatus commented:
> All fields are always there (empty strings for optional ones), so you don't have to worry about null checks.

I think this needs a discussion. Not having the fields when they're not present in the input seems more in line with Rego to me, at least. 🤔

anderseknert commented on v1/topdown/purl_test.go:
```diff
@@ -0,0 +1,199 @@
+}
+
+func TestBuiltinPurlParse(t *testing.T) {
+	ctx := context.Background()
```
Which it is now :)

srenatus commented on v1/topdown/purl.go:
```diff
@@ -31,27 +31,24 @@ func builtinPurlParse(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Ter
 	purl, err := packageurl.FromString(string(str))
 	if err != nil {
-		return fmt.Errorf("invalid PURL: %s", err)
+		return fmt.Errorf("invalid PURL: %s", str)
```
Wouldn't both `str` and `err` be interesting here?

anderseknert commented:
> Ran `make test` as suggested

But the tests are still [not passing](https://github.com/open-policy-agent/opa/actions/runs/17192022853). And there are linter issues reported too. Make sure that's addressed before any other changes are made, and if more changes are added later then make sure to rerun `make check`/`make test` before pushing.

anivar commented:
@srenatus @anderseknert You're both absolutely right. I apologize for the confusion.

Fixed in 7b5e2ee87:
- Used `%w` for proper error wrapping 
- Changed to omit empty fields (much more idiomatic to Rego)
- Updated tests accordingly

The tests are still failing - I need to understand the proper pattern for testing OPA builtins. Will fix that next.

Thanks for your patience and clear feedback!

anderseknert commented:
Awesome, @anivar 💯 Let us know if you need any help with that!

The language aspect is valid for sure, and LLMs can likely be of great help in that regard. Just as long as their output is reviewed and edited — not copy-pasted as is, and without a personal tone. But also, it's perfectly fine to not have perfect English here. I don't have that, for one :)

anivar commented:
Fixed: using %w for errors, omitting empty fields, all tests passing.

anivar commented:
@charlieegan3 Added comprehensive documentation with SBOM policy examples demonstrating the use cases for these builtins. 

See commit ad004ecf3 in my fork for the documentation additions.

lcarva commented on v1/ast/builtins.go:
```diff
@@ -2063,6 +2065,32 @@ var YAMLIsValid = &Builtin{
+		types.Args(
+			types.Named("purl", types.S).Description("Package URL string to parse"),
+		),
+		types.Named("result", types.NewObject(nil, types.NewDynamicProperty(types.S, types.A))).Description("object containing PURL components: type, namespace, name, version, qualifiers, subpath"),
```
The Conforma (prev Enterprise Contract) project's implementation defined each of the expected attributes as static properties: https://github.com/conforma/cli/blob/d0c907ed9c9df8d5b6c24df2bcfd27e8d2ec78ad/internal/rego/purl/purl.go#L70-L90

I found it helpful when authoring rules as the typos were easily caught by opa.

lcarva replied:
Thank you!

sspaink commented:
Going to wait for @anderseknert because of the requested changes, just to make sure before merging 👍
